contracts/marketplace/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+#![no_std]
+use soroban_sdk::{
+    contract, contracterror, contractimpl, contracttype, panic_with_error,
+    token::{self},
+    Address, Env, Map, Symbol,
+};
+
+#[derive(Clone, Copy)]
+#[contracttype]
+enum DataKey {
+    Admin = 1,
+    Percentage = 2,
+    AlreadyInitialized = 3,
+    Assets = 4,
+    Token = 5,
+}
+
+#[contracterror]
+#[derive(Clone, Debug, Copy, Eq, PartialEq, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum Error {
+    AlreadyInitialized = 1,
+    InvalidPercentage = 2,
+    InvalidAssetPrice = 3,
+    BalanceTooLow = 4,
+    AssetNotListed = 5,
+    InvalidAuth = 6,
+}
+
+#[contracttype]
+#[derive(Clone)]
+pub struct Asset {
+    owner: Address,
+    price: i128,
+    listed: bool,
+}
+
+#[contract]
+pub struct MarketplaceContract;
+
+#[contractimpl]
+impl MarketplaceContract {
+    /// Initialize the contract with the admin's address and a percentage for the admin fee.
+    pub fn init(env: Env, token: Address, admin: Address, percentage: i128) {
+        admin.require_auth();
+        let storage = env.storage().persistent();
+
+        if storage
+            .get::<_, bool>(&DataKey::AlreadyInitialized)
+            .is_some()
+        {
+            panic_with_error!(&env, Error::AlreadyInitialized);
+        }
+
+        if !(0..100).contains(&percentage) {
+            panic_with_error!(&env, Error::InvalidPercentage);
+        }
+        storage.set(&DataKey::Admin, &admin);
+        storage.set(&DataKey::Token, &token);
+        storage.set(&DataKey::Percentage, &percentage);
```

Comment:
This is stored, but when its used ? Is it the idea to retain a sales fee for the the transaction for the admin of the contract ?

Reply:
Yes, the admin would set on init.

Reply:
Hmm, it should during buy listing

Reply:
I am actually wondering if we should remove the part for percentages.
Leave that to anyone working reusing the contract.

Reply:
Not sure. I would involve here @mariopil also. 

My thoughts: 

If this wasnt a smart contract, i would say to implement it. Theres no marketplace out there that doesnt charge fees on their global sales. But taken into account this is a smart contract, things could bedifferent.

If we are in doubt we can always add it. Then set it to zero depending on the use case.

Reply:
Done